sql/src/main/java/io/crate/analyze/InsertAnalyzer.java

```diff
@@ -139,12 +143,53 @@ public AnalyzedInsertStatement analyze(Insert<?> insert, ParamTypeHints typeHint
         final boolean ignoreDuplicateKeys =
             insert.duplicateKeyContext().getType() == Insert.DuplicateKeyContext.Type.ON_CONFLICT_DO_NOTHING;
 
+        List<Symbol> returnValues = List.of();
+        List<ColumnIdent> outputNames = List.of();
+
+        if (!insert.returningClause().isEmpty()) {
+            var stmtCtx = new StatementAnalysisContext(typeHints, Operation.INSERT, txnCtx);
+            var relCtx = stmtCtx.startRelation();
+            var relation = relationAnalyzer.analyze(insert.table(), stmtCtx);
+            stmtCtx.endRelation();
+
+            var exprCtx = new ExpressionAnalysisContext();
+            var sourceExprAnalyzer = new ExpressionAnalyzer(
+                functions,
+                txnCtx,
+                typeHints,
+                new FullQualifiedNameFieldProvider(
+                    relCtx.sources(),
+                    relCtx.parentSources(),
+                    txnCtx.sessionContext().searchPath().currentSchema()
+                ),
+                null
+            );
+
+            var selectAnalysis = SelectAnalyzer.analyzeSelectItems(
+                insert.returningClause(),
+                relCtx.sources(),
+                sourceExprAnalyzer,
+                exprCtx
+            );
+
+            var normalizer = new EvaluatingNormalizer(functions,
+                                                      RowGranularity.CLUSTER,
+                                                      null,
+                                                      (FieldResolver) relation);
+
+            returnValues = Lists2.map(selectAnalysis.outputSymbols(), x -> normalizer.normalize(x, txnCtx));
```

Comment:
I'm not sure if we should do the normalization here already.

This is to turn `Field` into `Reference` I assume?

In other cases we do this as part of the logical planning via the `RelationNormalizer`. Maybe you could have a look if it would be possible to also defer this to the `RelationNormalizer`.

I wouldn't spend too much time on it. I want to change some things about `Field` and `Reference` soon anyhow.

Reply:
CurrentIy we normalize also in the `UpdateAnalyzer` at the same point. I could extend the `RelationNormalizer` to also support the `AnalyzedInsertStatement`. Does this make sense to you ?

Reply:
> could extend the RelationNormalizer to also support the AnalyzedInsertStatement. Does this make sense to you ?

Yes, but only if it isn't too much effort and doesn't cause other issues.

Reply:
I had a look at `RelationNormalizer` and I think conceptionally it would be a bit awkward to add  `AnalyzedInsertStatement` as a non-relational type to the `RelationNormalizer`. I would leave as it is right now, what do you think ?